File: python/reference/splay.py
"""Bottom-up Splay, exactly the five frozen cases (reference implementation).

Cases: ZIG, ZIG-ZIG LL, ZIG-ZIG RR, ZIG-ZAG LR, ZIG-ZAG RL.
Post-condition of splay: x is root, tree is a valid BST on the same keys.
"""

from .tree import EMPTY, find_path, replace_node, subtree, tree_size, validate_bst


def rotate_right(t, p):
    """Right rotation at node p. Requires a left child (else ValueError)."""
    path = find_path(t, p)
    left, pk, right = subtree(t, p)
    if left == ():
        raise ValueError("rotate_right without left child at %r" % (p,))
    ll, qk, lr = left
    new_sub = (ll, qk, (lr, pk, right))
    return replace_node(t, path, new_sub)


def rotate_left(t, p):
    """Left rotation at node p. Requires a right child (else ValueError)."""
    path = find_path(t, p)
    left, pk, right = subtree(t, p)
    if right == ():
        raise ValueError("rotate_left without right child at %r" % (p,))
    rl, qk, rr = right
    new_sub = ((left, pk, rl), qk, rr)
    return replace_node(t, path, new_sub)


def splay(t, x):
    """Bottom-up splay of x. Returns (new_tree, cost, cases, access_path)."""
    n = tree_size(t)
    path0 = find_path(t, x)
    cost = len(path0)
    cases = []
    cur = t
    while True:
        path = find_path(cur, x)
        if len(path) == 1:
            break
        v = path[-1]
        p = path[-2]
        v_is_left = v < p
        if len(path) == 2:
            cur = rotate_right(cur, p) if v_is_left else rotate_left(cur, p)
            cases.append("ZIG")
        else:
            g = path[-3]
            p_is_left = p < g
            if v_is_left and p_is_left:
                cur = rotate_right(cur, g)
                cur = rotate_right(cur, p)
                cases.append("LL")
            elif not v_is_left and not p_is_left:
                cur = rotate_left(cur, g)
                cur = rotate_left(cur, p)
                cases.append("RR")
            elif not v_is_left and p_is_left:
                cur = rotate_left(cur, p)
                cur = rotate_right(cur, g)
                cases.append("LR")
            else:
                cur = rotate_right(cur, p)
                cur = rotate_left(cur, g)
                cases.append("RL")
    if find_path(cur, x) != [x]:
        raise AssertionError("x is not root after splay")
    if not validate_bst(cur, n):
        raise AssertionError("BST invalid after splay")
    return cur, cost, cases, path0

File: crates/splay_model/src/splay.rs
//! Bottom-up Splay: exactly ZIG / LL / RR / LR / RL (frozen spec Sec. 4).

use crate::rotate::{child_side, rotate_left, rotate_right};
use crate::tree::{Tree, TreeError};

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct SplayOutcome {
    pub cost: usize,
    pub cases: Vec<&'static str>,
    pub access_path: Vec<u32>,
}

/// Splay key x to the root. Returns pre-splay cost, rotation cases, and the
/// pre-splay access path. Post-conditions: x is root, tree checks clean.
pub fn splay(t: &mut Tree, x: u32) -> Result<SplayOutcome, TreeError> {
    let target = t.index_of(x).ok_or(TreeError::KeyAbsent(x))?;
    let _ = target;
    let path: Vec<u32> = {
        let mut v = t.index_of(x).unwrap();
        let mut rev = vec![t.nodes[v].key];
        while let Some(p) = t.nodes[v].parent {
            v = p;
            rev.push(t.nodes[v].key);
        }
        rev.reverse();
        rev
    };
    let cost = path.len();
    let mut cases = Vec::new();
    loop {
        let vi = t.index_of(x).unwrap();
        let pi = match t.nodes[vi].parent {
            None => break,
            Some(p) => p,
        };
        let v_left = child_side(t, vi)?.unwrap_or(true);
        match t.nodes[pi].parent {
            None => {
                if v_left {
                    rotate_right(t, pi)?;
                } else {
                    rotate_left(t, pi)?;
                }
                cases.push("ZIG");
            }
            Some(gi) => {
                let p_left = child_side(t, pi)?.unwrap_or(true);
                match (v_left, p_left) {
                    (true, true) => {
                        rotate_right(t, gi)?;
                        rotate_right(t, pi)?;
                        cases.push("LL");
                    }
                    (false, false) => {
                        rotate_left(t, gi)?;
                        rotate_left(t, pi)?;
                        cases.push("RR");
                    }
                    (false, true) => {
                        rotate_left(t, pi)?;
                        rotate_right(t, gi)?;
                        cases.push("LR");
                    }
                    (true, false) => {
                        rotate_right(t, pi)?;
                        rotate_left(t, gi)?;
                        cases.push("RL");
                    }
                }
            }
        }
    }
    t.check()?;
    let root_key = t.nodes[t.root.unwrap()].key;
    if root_key != x {
        return Err(TreeError::Invariant("x not root after splay".to_string()));
    }
    Ok(SplayOutcome { cost, cases, access_path: path })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::encode::parse_shape;
    use crate::tree::Tree;

    fn shape_of(t: &Tree) -> String {
        fn go(t: &Tree, idx: Option<usize>) -> String {
            match idx {
                None => ".".to_string(),
                Some(i) => format!("({}{})", go(t, t.nodes[i].left), go(t, t.nodes[i].right)),
            }
        }
        go(t, t.root)
    }

    #[test]
    fn zigzig_left_chain_matches_fixture() {
        let mut t = Tree::from_shape(&parse_shape("(((..).).)").unwrap());
        let out = splay(&mut t, 1).unwrap();
        assert_eq!(out.cost, 3);
        assert_eq!(out.cases, vec!["LL", "ZIG"]);
        assert_eq!(shape_of(&t), "(.(.(..)))");
    }

    #[test]
    fn zigzag_left_right_matches_fixture() {
        let mut t = Tree::from_shape(&parse_shape("((.(..)).)").unwrap());
        let out = splay(&mut t, 2).unwrap();
        assert_eq!(out.cost, 3);
        assert_eq!(out.cases, vec!["LR"]);
        assert_eq!(shape_of(&t), "((..)(..))");
    }

    #[test]
    fn root_access_is_identity() {
        let mut t = Tree::from_shape(&parse_shape("((..)(..))").unwrap());
        let out = splay(&mut t, 2).unwrap();
        assert_eq!(out.cost, 1);
        assert!(out.cases.is_empty());
        assert_eq!(shape_of(&t), "((..)(..))");
    }
}
